flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java

```diff
@@ -879,6 +875,42 @@ else if (transitionState(current, ExecutionState.FAILED, t)) {
 		}
 	}
 
+	/**
+	 * There are two scenarios to release the network resources. One is from {@link TaskCanceler} to early
+	 * release partitions and gates. Another is from task thread during task exiting.
+	 */
+	private void closeNetworkResources(boolean isCanceling) {
```

Comment:
I would break this method down:
```
private void releaseNetworkResources() {
  for all partitions taskEventDispatcher.unregisterPartition
  Throwable cause = isCanceledOrFailed() ? getFailureCause() : null;
  closeOrFailNetworkResources(producedPartitions, inputGates, cause);
}

static closeNetworkResources(producedPartitions, inputGates, cause) {
  // closing loops or fail for partition if cause is null
}

TaskCanceler.run() {
  closeOrFailNetworkResources(producedPartitions, inputGates, null); // to preserve what we have
}
```
I would also keep TaskCanceler class static to simplify refactoring in future.

Reply:
I agree with the option of keeping previous static TaskCanceler class. 
But the static `closeNetworkResources` might need two more parameters. One is for `taskNameWithSubtask` used for log, and another is the boolean `isCanceledOrFailed` because in previous behavior the result of  `isCanceledOrFailed` is not always equal to `cause != null`. In the case of `Task#cancelExecution`, the cause is null but `isCanceledOrFailed` would return true. What do you think?